big_tests/tests/sm_SUITE.erl

```diff
@@ -793,6 +803,117 @@ wait_for_resumption(Config) ->
     assert_no_offline_msgs(AliceSpec),
     wait_for_c2s_state_change(C2SPid, resume_session).
 
+unacknowledged_message_hook_resume(Config) ->
+    unacknowledged_message_hook(?FUNCTION_NAME,Config).
+
+unacknowledged_message_hook_bounce(Config) ->
+    unacknowledged_message_hook(?FUNCTION_NAME, Config).
+
+unacknowledged_message_hook_offline(Config) ->
+    unacknowledged_message_hook(?FUNCTION_NAME, Config).
+
+unacknowledged_message_hook(TestCase, Config) ->
+    ConnSteps = connection_steps_to_session(),
+
+    %% connect bob and alice
+    BobSpec = escalus_fresh:create_fresh_user(Config, bob),
+    {ok, Bob, _} = escalus_connection:start(BobSpec),
+    escalus_connection:send(Bob, escalus_stanza:presence(<<"available">>)),
+    escalus_connection:get_stanza(Bob, presence),
+
+    AliceSpec0 = escalus_fresh:create_fresh_user(Config, alice),
+    Resource = proplists:get_value(username, AliceSpec0),
+    AliceSpec = [{resource, Resource} | AliceSpec0],
+    start_hook_listener(Resource),
+    {ok, Alice, _} = escalus_connection:start(AliceSpec, ConnSteps ++ [stream_resumption]),
+    SMID = proplists:get_value(smid, Alice#client.props),
+    escalus_connection:send(Alice, escalus_stanza:presence(<<"available">>)),
+    escalus_connection:get_stanza(Alice, presence),
+    %% Ack the presence stanza
+    escalus:assert(is_sm_ack_request, escalus_connection:get_stanza(Alice, ack)),
+    escalus:send(Alice, escalus_stanza:sm_ack(1)),
+
+    escalus_connection:send(Bob, escalus_stanza:chat_to(common_helper:get_bjid(AliceSpec), <<"msg-1">>)),
+    escalus_connection:send(Bob, escalus_stanza:chat_to(common_helper:get_bjid(AliceSpec), <<"msg-2">>)),
+    %% kill alice connection
+    {ok, C2SPid} = get_session_pid(AliceSpec, Resource),
+    escalus_connection:kill(Alice),
+    wait_for_c2s_state_change(C2SPid, resume_session),
+    1 = length(get_user_alive_resources(AliceSpec)),
+
+    escalus:assert(is_chat_message, [<<"msg-1">>], verify_hook(0, Resource, 100)),
+    escalus:assert(is_chat_message, [<<"msg-2">>], verify_hook(0, Resource, 100)),
+    ?assertEqual(timeout, verify_hook(0, Resource, 100)),
+
+    %% send some messages and check if c2s can handle it
+    escalus_connection:send(Bob, escalus_stanza:chat_to(common_helper:get_bjid(AliceSpec), <<"msg-3">>)),
+    escalus_connection:send(Bob, escalus_stanza:chat_to(common_helper:get_bjid(AliceSpec), <<"msg-4">>)),
+    escalus:assert(is_chat_message, [<<"msg-3">>], verify_hook(0, Resource, 100)),
+    escalus:assert(is_chat_message, [<<"msg-4">>], verify_hook(0, Resource, 100)),
+    ?assertEqual(timeout, verify_hook(0, Resource, 100)),
+
+    %% alice comes back and receives unacked message
+    {NewResource, NewAlice} =
+        case TestCase of
+            unacknowledged_message_hook_resume ->
```

Comment:
I think that the current structure is complicated and not very flexible. It would be more practical to have the specific case include:

```
specific(Config) ->
  SomeRes = common1(Config),
  SomeRes2 = specific_logic(SomeRes),
  common2(Config, SomeRes, SomeRes)
```

But now we have a flow where specific immediately calls common helper, which calls specific logic based on the parent name. More difficult to reason about.

Reply:
the problem is that common1 part has to return 5 parameters. and that is not very readable as well.